composer/datasets/streaming/format.py

```diff
@@ -0,0 +1,227 @@
+    for key in keys:
+        value = obj[key]
+        values.append(value)
+    sizes = list(map(len, values))
+    sizes = np.array(sizes, np.int32)
+    return sizes.tobytes() + b''.join(values)
+
+
+def bytes_to_sample_dict(data: bytes, keys: Sequence[str]) -> Dict[str, bytes]:
+    """Load a sample dict from bytes and field names.
+
+    Args:
+        data (bytes): The encoded sample data.
+        keys (list of str): The field names.
+
+    Returns:
+        dict: The decoded sample dict.
+    """
+    num_values = len(keys)
+    sizes = np.frombuffer(data[:num_values * 4], np.int32)
+    ends = num_values * 4 + sizes.cumsum(0)
+    begins = ends - sizes
+    values = []
+    for begin, end in zip(begins, ends):
+        value = data[begin:end]
+        values.append(value)
+    return dict(zip(keys, values))
+
+
+class StreamingDatasetIndex(object):
+    """Streaming dataset index file, containing all the info about shards.
+
+    The shards are just dumb buffers with samples catted together. All the
+    offset info across the whole dataset is contained in the index file. Workers
+    read this file to calculate how much of which shards their slice is.
+
+    Each sample is a dict of str to bytes. All samples must contain the same
+    dict keys (fields). These strings are stored in the index file for efficiency.
+    """
+
+    def __init__(self, samples_per_shard: Sequence[int], bytes_per_shard: Sequence[int],
+                 bytes_per_sample: Sequence[int], fields: Sequence[str]) -> None:
+        """Initialize with sample statistics.
+
+        Args:
+            samples_per_shard (Sequence[int]): Number of samples of each shard.
+            bytes_per_shard (Sequence[int]): Size in bytes of each shard.
+            bytes_per_sample (Sequence[int]): Size in bytes of each sample across all shards.
+            fields (Sequence[str]): The names of the samples' fields in order.
+        """
+        self.samples_per_shard = np.asarray(samples_per_shard, np.int32)
+        self.bytes_per_shard = np.asarray(bytes_per_shard, np.int32)
+        self.bytes_per_sample = np.asarray(bytes_per_sample, np.int32)
+        self.fields = fields
+
+        # Totals.
+        self.num_shards = len(samples_per_shard)
+        self.total_bytes = sum(bytes_per_shard)
+        self.total_samples = len(bytes_per_sample)
+        self.num_fields = len(fields)
+
+        # Shard -> sample range.
+        self.shard_ends = self.samples_per_shard.cumsum()
+        self.shard_begins = self.shard_ends - self.samples_per_shard
+
+        # Sample -> shard, byte offset within shard.
+        self.sample_shards, self.sample_shard_offsets = self.locate_samples()
+
+    @classmethod
+    def loads(cls, data: bytes):
+        """Load a StreamingDatasetIndex from raw bytes.
+
+        Args:
+            data (bytes): The serialized form.
+
+        Returns:
+            cls: The loaded object.
+        """
+        begin = 0
+        end = begin + 2 * 8
```

Comment:
These constants (2, 8) seem a bit magical. Could you define them somewhere or read them from a library

Reply:
Done